src/plugins/data_source/server/plugin.ts

```diff
@@ -3,16 +3,28 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-import { PluginInitializerContext, CoreSetup, CoreStart, Plugin, Logger } from 'src/core/server';
+import {
+  PluginInitializerContext,
+  CoreSetup,
+  CoreStart,
+  Plugin,
+  Logger,
+  IContextProvider,
+  RequestHandler,
+} from '../../../../src/core/server';
+import { DataSourceClient } from './client/data_source_client';
+import { DataSourceRouteHandlerContext } from './data_source_route_handler_context';
 import { dataSource, credential } from './saved_objects';
 
 import { DataSourcePluginSetup, DataSourcePluginStart } from './types';
 
 export class DataSourcePlugin implements Plugin<DataSourcePluginSetup, DataSourcePluginStart> {
   private readonly logger: Logger;
+  private readonly dataSourceClient: DataSourceClient;
 
   constructor(initializerContext: PluginInitializerContext) {
     this.logger = initializerContext.logger.get();
+    this.dataSourceClient = new DataSourceClient(this.logger);
```

Comment:
by having one single `DataSourceClient` object, and you may update the saved object client in that object for each request, will it causing racing condition and mess up the DataSourceClient object internal state?

Reply:
I am actually not sure, cuz I thought node.js is single threaded, and safe. But anyways, I added another layer of abstraction "dataSourceService", which has function to create `dataSourceClient` with `scopedSavedObjectClient` on demand,  could you take another look? I feel that makes more sense

fyi @noCharger

Reply:
NodeJs is single threaded, but the request processing is async, so you still have the issue when your class internal state is messed up